### DISCUSSION TITLE
Adding more packages to the dockerfile for extensions?

### DISCUSSION THREAD

**Discussion by kelvie** [AbdBarho/stable-diffusion-webui-docker#453] [Ideas]
I just spent some time debugging why the dreambooth (in auto) doesn't work right anymore, and finally narrowed it down to the fact that one of the modules, bitsandbytes, requires a bunch of cuda libs: https://github.com/TimDettmers/bitsandbytes/issues/85

```
# ldd /usr/local/lib/python3.10/site-packages/bitsandbytes/libbitsandbytes_cuda118.so
        linux-vdso.so.1 (0x00007ffeeeb7c000)
        libtcmalloc.so => /usr/lib/x86_64-linux-gnu/libtcmalloc.so (0x00007f47ae600000)
        libcudart.so.11.0 => /usr/lib/x86_64-linux-gnu/libcudart.so.11.0 (0x00007f47ae200000)
        libcublas.so.11 => /usr/lib/x86_64-linux-gnu/libcublas.so.11 (0x00007f47a8400000)
        libcublasLt.so.11 => /usr/lib/x86_64-linux-gnu/libcublasLt.so.11 (0x00007f4783e00000)
        libcurand.so.10 => /usr/local/cuda/targets/x86_64-linux/lib/libcurand.so.10 (0x00007f477d400000)
        libcusparse.so.11 => /usr/local/cuda/targets/x86_64-linux/lib/libcusparse.so.11 (0x00007f476c600000)
        libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f47ae833000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f47ae4bc000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f47af52a000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f476c42c000)
        libunwind.so.8 => /usr/lib/x86_64-linux-gnu/libunwind.so.8 (0x00007f47ae818000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f47ae1de000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f47af54f000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f47af522000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f47ae80e000)
        liblzma.so.5 => /lib/x86_64-linux-gnu/liblzma.so.5 (0x00007f47ae1b6000)
```

Being that nvidia and python are both atrocious for packaging, I totally get it, but perhaps this might not be the only other python library that expects to be able to find cuda libs. Should this docker image consider adding (back?) cuda-libraries-dev? 

Otherwise I'm just going to be maintaining my own patches to the dockerfile on top of this one to let some extensions work, currently, this is what I add to my dockerfile to get my set of extensions working:
```
# For some other extensions
RUN aria2c https://developer.download.nvidia.com/compute/cuda/repos/debian11/x86_64/cuda-keyring_1.0-1_all.deb
RUN dpkg -i cuda-keyring_1.0-1_all.deb
RUN apt-get update && apt-get -y install build-essential libmariadb-dev pkg-config libcairo2-dev cuda-libraries-dev-11-8 && apt-get clean
``` 
Is there a cleaner and more maintanable way to add packages I need to the dockerfile? It was mentioned that the `startup.sh` is the best way for doing the pip installs, but it takes a lot longer to do the apt-installs every time, and it seems harder to cache (the pip cache is persisted across docker compose runs)

**Comment by kelvie** [comment]
https://github.com/d8ahazard/sd_dreambooth_extension/issues/1023

**Reply by AbdBarho** [reply]
On top of my head one thing you can try is to switch the base image from `python-slim` to one of the [pytorch containers](https://hub.docker.com/layers/pytorch/pytorch/2.0.0-cuda11.7-cudnn8-runtime/images/sha256-fabb5a665a05b8ee0ac76f0d943acc40039e13536e11a44d3dc47625a266e759?context=explore). These have all cuda stuff (at least in theory).

I remember deciding against this because these containers have python 3.7 or 3.8 per default, but most UIs need 3.10. Not sure what the state is now.

**Comment by chakflying** [comment]
I also found this when installing the controlnet extension, I needed to add `build-essential libcairo2-dev pkg-config` to the `Dockerfile`. IMO the container should include these common packages so people have less trouble when installing extensions.

Regarding the PyTorch images, yes they are still using Python 3.8, we can track this here:
https://docs.nvidia.com/deeplearning/frameworks/pytorch-release-notes/rel-23-04.html

**Reply by AbdBarho** [reply]
Yeah, I agree that we need some shared dependencies already installed in the container. The question is, where do we stop? how much is enough?

**Reply by kelvie** [reply]
Maybe we can have an "EXTRA_APT_PACKAGES" ARG in the dockerfile that we can specify in our own docker-compose files, with a list of extra packages to install as part of the dockerfile build.

Cuda is a bit more complicated, as it requires downloading of a deb, installing it, then an `apt upgrade` and `apt install`,  so the naïve suggestion would be an `EXTRA_APT_REPOSITORY_DEB_URLS`, but now it's just getting crazy.

**Reply by chakflying** [reply]
I'm not a fan of more configuration, since the main benefit of using docker is that things work out of the box. But I understand there is a concern of the image getting bloated.

Here is the official list of extensions sorted by the number of stars, top 10:
```
Mikubill/sd-webui-controlnet:                               8025
pkuliyi2015/multidiffusion-upscaler-for-automatic1111:      2046
DominikDoom/a1111-sd-webui-tagcomplete:                     1420
d8ahazard/sd_dreambooth_extension:                          1266
deforum-art/deforum-for-automatic1111-webui:                1250
fkunn1326/openpose-editor:                                  1116
kohya-ss/sd-webui-additional-networks:                      1018
thygate/stable-diffusion-webui-depthmap-script:              894
adieyal/sd-dynamic-prompts:                                  761
dtlnor/stable-diffusion-webui-localization-zh_CN:            761
```

If you want to be qualitative, you can come up with some metric base on this. But I would say it's not excessive to support the top extension by default.

**Reply by AbdBarho** [reply]
@chakflying sorry for late answer.

I think supporting top extensions should be fine, do you know which `apt` packages are needed for these extensions? maybe we can start with the top one or two?

It would be great if you could create a PR.

**Comment by conor-is-my-name** [comment]
I agree that getting extensions dependencies is the biggest pain point right now.  I am a relatively amateur programmer so please take my proposal with a grain of salt.

Up until now I have been manually adding lines to the docker file similar to @kelvie method above.  However, it's a tedious process.

Is it possible to wrap the whole thing in 1 virtual environment? If this were the case, I believe that the launch import that is present in the install.py files of the extensions would be able to call the launch.py file in the automatic1111 repo.  If this is possible, I think it might be a better solution than maintaining a separate hypothetical requirements.txt file for common packages.

**Comment by DevilaN** [comment]
@cbolich:
Why do you want to mess with image when `startup.sh` can be used to install dependencies for extensions?

**Reply by conor-is-my-name** [reply]
I still have to install many dependencies separately.  For what it's worth here are my additions to the dockerfile:

RUN apt-get update && apt-get install -y \
    libglvnd0 \
    libgl1 \
    libglx0 \
    libegl1 \
    libglvnd-dev \
    libgl1-mesa-dev \
    libegl1-mesa-dev \
    libxext6 \
    libx11-6 \
    libegl1-mesa \
    xorg-dev \
    libgl1-mesa-glx \
    libgl1-mesa-dri \
    libglfw3 \
    mesa-utils \
    fontconfig \
    ffmpeg

ENV NVIDIA_VISIBLE_DEVICES all
ENV NVIDIA_DRIVER_CAPABILITIES graphics,utility,compute

RUN pip install vispy
RUN pip install trimesh
RUN pip install moviepy
RUN pip install --upgrade tqdm
RUN pip install transforms3d
RUN pip install rembg
RUN pip install imageio_ffmpeg


update:
I want to add that I am still having trouble getting all my extensions to work.  Keep getting EGL errors

**Reply by DevilaN** [reply]
Instead of changing Dockerfile you should do it in `startup.sh`.

**Comment by AbdBarho** [comment]
> I believe that the launch import that is present in the install.py files of the extensions would be able to call the launch.py file in the automatic1111 repo.

You can call any scripts you want in the `startup.sh`

> I want to add that I am still having trouble getting all my extensions to work. Keep getting EGL errors

Yeah, and unfortunately, I don't see that getting better anytime soon. Every extension has its own set of dependencies and they overlap and overwrite each other.

I don't think this is a problem that we can solve in this repo, we may make it a bit easier by installing a "separate hypothetical requirements.txt file for common packages", but this is an ecosystem problem.

**Reply by DevilaN** [reply]
I've got a feeling that this subject pops up too often. Should we just add info about this in FAQ?

**Reply by AbdBarho** [reply]
Sure! do you have a suggestion in mind? or should I just improvise 😅

**Reply by DevilaN** [reply]
@AbdBarho It's already there. Most of it. https://github.com/AbdBarho/stable-diffusion-webui-docker/wiki/Usage#extensions
Probably write in CAPS with red letters. Blinking text will probably also have positive effect xD

I would aim for something like this (but You should rephrase it probably).
Different extensions require additional dependencies.  Some of them might conflict with each other and changing versions of packages could break things. To make sure everything out of the box is working fine you should never modify Dockerfile / make changes to image. Instead use `startup.sh` script to install dependencies needed. In case of something goes wrong you can always comment section in `startup.sh` file which install your additional dependencies, and do `docker compose --profile auto down` to clean up messy containers, so you'll start with clean and working container.

**Reply by AbdBarho** [reply]
Thanks, updated!

**Comment by dereekb** [comment]
Ran into this tonight. I'm using this repo to setup ComfyUI and was having trouble installing the [WAS Node Suite](https://github.com/WASasquatch/was-node-suite-comfyui) since it requires dlib, and requires a couple dependencies from apt-get to properly build/install the dlib python library/dependency.

`apt-get install -y cmake g++ libprotobuf-dev protobuf-compiler`

I ended up just making a shell script that I run to update the ComfyUI Dockerfile (`/services/comfy`), and one step is to replace line 12 of the Dockerfile with the dependencies using sed.

`sed -i '12s/.*/RUN apt-get update \&\& apt-get install -y git cmake g++ libprotobuf-dev protobuf-compiler \&\& apt-get clean/' Dockerfile`

I'm happy enough with that solution for now, but it would be nice to have dlib pre-installed for the ComfyUI setup, since it's one of the main recommended plugins.

(Full script for reference)

`comfy.sh`
```
cd ./stable-diffusion-webui-docker/services/comfy
# Replace the dependencies install line
sed -i '12s/.*/RUN apt-get update \&\& apt-get install -y git cmake g++ libprotobuf-dev protobuf-compiler \&\& apt-get clean/' Dockerfile
# Install Custom Nodes
cd ../../data/config/comfy/custom_nodes
git clone https://github.com/ltdrdata/ComfyUI-Manager.git
git clone https://github.com/WASasquatch/was-node-suite-comfyui/
cd ../../../../
docker compose --profile comfy up --build
```

**Reply by DevilaN** [reply]
Please don't do that! My eyes are bleeding when I see it.
Can't you just leave original image as it is? There are better and less hacky options. You can derive image from original one and add your layer there. Then you can use `docker-compose.override.yml` to force your image instead of original one.

**Reply by dereekb** [reply]
> Please don't do that! My eyes are bleeding when I see it. Can't you just leave original image as it is? There are better and less hacky options. You can derive image from original one and add your layer there. Then you can use `docker-compose.override.yml` to force your image instead of original one.

You mean build the first one, then use the FROM command to extend the new one? I was looking at doing that at first, but didn't want to depend on building the first image and then building another one. I find myself rebuilding the Comfy image enough that just making edits to the original Dockerfile seemed fine.

The only issue is doubting that the apt-get line is going to get updated/touched. This way is kept outside of the folder and is safe between git pulls and I can blow away the folder without worrying about losing these scripts.

The image itself gets rebuilt as part of the `docker compose --profile comfy up --build`, so I don't particularly see that there are better and less hacky options that are as easy as the above.

Is the `startup.sh` usage part of the ComfyUI build? I'd just do that (although it would forgo Docker's caching), but it wasn't clear in the wiki that was a possibility.

**Reply by DevilaN** [reply]
@dereekb: Yes, that's what I mean. There are several drawbacks when you edit Dockerfile. When using docker you want to benefit from all of it's pros and not to limit your flexibility with images. Right now your method is somewhat easy to do when you mastered all "change that" stuff. But later you might encounter some problem and using modified image almost all of the time will make debugging harder.

You can emulate `startup.sh` with mounting file and changing entrypoint of service. But it would be better if @AbdBarho added proper functionality for this.

**Comment by simonmcnair** [comment]
as an aside I tried the advice from here https://github.com/AbdBarho/stable-diffusion-webui-docker/wiki/Usage#extensions

And got the following error:

/data/config/auto/startup.sh: line 2: shopt: nullglob
: invalid shell option name

**Reply by simonmcnair** [reply]
my bad.  Windows EOL rather than Linux.  Soz

**Comment by simonmcnair** [comment]
this is okay for me atm
```
# install all packages for the extensions
shopt -s nullglob
aria2c https://developer.download.nvidia.com/compute/cuda/repos/debian11/x86_64/cuda-keyring_1.0-1_all.deb
dpkg -i cuda-keyring_1.0-1_all.deb
apt-get update && apt-get -y install build-essential libmariadb-dev pkg-config libcairo2-dev cuda-libraries-dev-11-8 && apt-get clean
pip install --upgrade pip

pip install ffmpeg diffusers discord_webhook
#discord_webhook screws A1111's clean-fid 0.1.29 requires requests==2.25.1, but you have requests 2.31.0 which is incompatible.
#Still starts up though.  dreambooth will not start without discord_webhook though.
pip install -q --force-reinstall opencv-python-headless
```

**Comment by simonmcnair** [comment]
I have installed all the extensions from
https://github.com/AbdBarho/stable-diffusion-webui-docker/discussions/453#discussioncomment-5857274

With the modifications from
https://github.com/AbdBarho/stable-diffusion-webui-docker/discussions/453#discussioncomment-5976391

They all worked except for the following impacts:
Dreambooth
discord_webhook screws clean-fid 0.1.29 requires requests==2.25.1, but you have requests 2.31.0 which is incompatible.
You must install discord webhook or the extension does not initialise.  It all seems to work but cleanfid==0.1.29  is required by A1111

adieyal/sd-dynamic-prompt
imagereward send2trash dynamicprompts

thygate/stable-diffusion-webui-depthmap-script 
requires  apt-get install libglfw3-dev libgles2-mesa-dev

**Comment by simonmcnair** [comment]
### depthmap:
apt-get install libglfw3-dev libgles2-mesa-dev
and
opencv-python==4.2.0.32
vispy==0.6.4
moviepy==1.0.2
transforms3d==0.3.1
networkx==2.3
cynetworkx
scikit-image

deforum:
numexpr
matplotlib
pandas
av
pims
imageio_ffmpeg
rich
gdown

### controlnet:
mediapipe
svglib
fvcore
torch>=1.2.0
torchvision>=0.4.0

### dreambooth:
accelerate~=0.18.0
bitsandbytes==0.35.4
dadaptation~=1.5.0
diffusers~=0.14.0
discord-webhook~=1.1.0
fastapi~=0.94.1
gitpython~=3.1.31
lion-pytorch~=0.0.7
Pillow~=9.4.0
tqdm~=4.64.1
transformers~=4.26.1  # > 4.26.x causes issues (db extension #1110)
diffusers>=0.10.0

- Tensor
tensorboard==2.12.0; sys_platform != 'darwin' or platform_machine != 'arm64'
tensorflow==2.12.0; sys_platform != 'darwin' or platform_machine != 'arm64'

- Tensor MacOS
- Note: Major and Minor versions of tensorboard & tensorflow must match
tensorboard==2.11.2; sys_platform == 'darwin' and platform_machine == 'arm64'
tensorflow-macos==2.11.0; sys_platform == 'darwin' and platform_machine == 'arm64'
tensorflow-metal==0.7.1; sys_platform == 'darwin' and platform_machine == 'arm64'


### dynamicprompts:
send2trash==1.8.0
dynamicprompts[attentiongrabber,magicprompt]==0.23.0

**Reply by conor-is-my-name** [reply]
@simonmcnair I'm right there with you and my list looks pretty similar.

suggest you change:

transforms3d==0.4.1

Have you been able to get the video generation to work in the depth extension?  That's the final thing I'm having trouble getting to work.  I've tried all the solutions I could find in those threads, but no luck so far.  

The rest of the extension runs fine, its just that last video piece.  

webui-docker-auto-1  | Traceback (most recent call last):
webui-docker-auto-1  |   File "/stable-diffusion-webui/modules/call_queue.py", line 57, in f
webui-docker-auto-1  |     res = list(func(*args, **kwargs))
webui-docker-auto-1  |   File "/stable-diffusion-webui/modules/call_queue.py", line 37, in f
webui-docker-auto-1  |     res = func(*args, **kwargs)
webui-docker-auto-1  |   File "/stable-diffusion-webui/extensions/stable-diffusion-webui-depthmap-script/scripts/depthmap.py", line 1171, in run_generate
webui-docker-auto-1  |     outputs, mesh_fi, meshsimple_fi = run_depthmap(
webui-docker-auto-1  |   File "/stable-diffusion-webui/extensions/stable-diffusion-webui-depthmap-script/scripts/depthmap.py", line 766, in run_depthmap
webui-docker-auto-1  |     mesh_fi = run_3dphoto(device, inpaint_imgs, inpaint_depths, inputnames, outpath, fnExt, vid_ssaa, inpaint_vids)      
webui-docker-auto-1  |   File "/stable-diffusion-webui/extensions/stable-diffusion-webui-depthmap-script/scripts/depthmap.py", line 931, in run_3dphoto
webui-docker-auto-1  |     run_3dphoto_videos(mesh_fi, basename, outpath, 300, 40,
webui-docker-auto-1  |   File "/stable-diffusion-webui/extensions/stable-diffusion-webui-depthmap-script/scripts/depthmap.py", line 959, in run_3dphoto_videos
webui-docker-auto-1  |     vispy.use(app='egl')
webui-docker-auto-1  |   File "/usr/local/lib/python3.10/site-packages/vispy/util/wrappers.py", line 93, in use
webui-docker-auto-1  |     use_app(app)
webui-docker-auto-1  |   File "/usr/local/lib/python3.10/site-packages/vispy/app/_default_app.py", line 47, in use_app
webui-docker-auto-1  |     default_app = Application(backend_name)
webui-docker-auto-1  |   File "/usr/local/lib/python3.10/site-packages/vispy/app/application.py", line 47, in __init__
webui-docker-auto-1  |     self._use(backend_name)
webui-docker-auto-1  |   File "/usr/local/lib/python3.10/site-packages/vispy/app/application.py", line 240, in _use
webui-docker-auto-1  |     raise RuntimeError(msg)
webui-docker-auto-1  | RuntimeError: Could not import backend "EGL":
webui-docker-auto-1  | Could not initialize

**Reply by simonmcnair** [reply]
Sorry for the delay.  I don't use it, I just installed it to try it.  Finally got it it working with the following:

aria2c https://developer.download.nvidia.com/compute/cuda/repos/debian11/x86_64/cuda-keyring_1.0-1_all.deb
dpkg -i cuda-keyring_1.0-1_all.deb
apt-get update && apt-get -y install build-essential libmariadb-dev pkg-config libcairo2-dev cuda-libraries-dev-11-8 && apt-get clean
pip install --upgrade pip

apt-get install libglfw3-dev libgles2-mesa-dev

pip install ffmpeg diffusers discord_webhook trimesh rich vispy numexpr moviepy==1.0.2 onnx onnxruntime networkx==2.5 imageio imageio_ffmpeg transforms3d==0.4.1  importlib-metadata importlib-metadata matplotlib numba==0.57.0 rembg 

pip install -q --force-reinstall opencv-python-headless

**Comment by AbdBarho** [comment]
I have a PR with some new dependencies, if you could also recommend some missing ones, at least for the controlnet extension. that would be great.

https://github.com/AbdBarho/stable-diffusion-webui-docker/pull/483

**Reply by simonmcnair** [reply]
Might be worth having 2 options a1111 minimal and a1111 full.  If people don't use extensions they could go minimal or deal with it themselves ?

**Reply by AbdBarho** [reply]
maybe? I guess most people who use a1111 are there for controlnet or dreambooth.

**Comment by simonmcnair** [comment]
My latest looks like this:
```aria2c https://developer.download.nvidia.com/compute/cuda/repos/debian11/x86_64/cuda-keyring_1.0-1_all.deb
dpkg -i cuda-keyring_1.0-1_all.deb
apt-get update && apt-get -y install build-essential libmariadb-dev pkg-config libcairo2-dev cuda-libraries-dev-11-8 && apt-get clean
pip install --upgrade pip

apt-get install libglfw3-dev libgles2-mesa-dev

pip install ffmpeg diffusers discord_webhook trimesh rich vispy numexpr moviepy==1.0.2 onnx onnxruntime networkx==2.5 imageio imageio_ffmpeg transforms3d==0.4.1  importlib-metadata importlib-metadata matplotlib numba==0.57.0 rembg 

pip install -q --force-reinstall opencv-python-headless```

**Comment by cloudaxes** [comment]
The underlying issue is that Auto1111 installs the required packages in the container upon installation of the extension, but the extension is persisted outside of the container.  If the container is ever rebuilt, the extension is still present, but its dependencies have been wiped away.  You either need a simple store in the Docker container to track what extensions have already had dependencies installed in that container or you just pip install all the requirements for all extant extensions on every startup.

@AbdBarho:
> The question is, where do we stop? 

The answer is to not hardcode requirements at all and just have something on startup check through all the existing extensions and make sure their dependencies are installed in the container.  I did something like that here: https://github.com/AbdBarho/stable-diffusion-webui-docker/commit/04620dcb14128cc7f9da81f98ea13e1852d9c92b

**Reply by AbdBarho** [reply]
this is all mentioned in the wiki
https://github.com/AbdBarho/stable-diffusion-webui-docker/wiki/Usage#extensions

**Reply by cloudaxes** [reply]
Thanks!  That's great!  I was unaware that you already had a place to put this.  It makes a lot more sense now.  What was your thought process behind not adding this default startup.sh for everyone?  Just too much individual variation?  

Some extensions have an install.py that needs to be run.  I suggest the startup script should look something more like:
```shell
# install all packages for the extensions
shopt -s nullglob
list=(./extensions/*/requirements.txt)
for req in "${list[@]}"; do
  pip install -q -r "$req"
done
list_of_install_scripts=(./extensions/*/install.py)
for install_script in "${list_of_install_scripts[@]}"; do
  export PYTHONPATH="/stable-diffusion-webui"
  python "$install_script"
done
#opencv-python-headless to not rely on opengl and drivers.
pip install -q --force-reinstall opencv-python-headless
```

**Reply by simonmcnair** [reply]
As far as I can see half the requirements of one, break the requirements of another.  The other point is that this is only required on container rebuild, not restart, which can add a lot of start up delays.  As well as apt prospectively updating components out of band.

**Reply by AbdBarho** [reply]
> As far as I can see half the requirements of one, break the requirements of another.

Bingo

> The other point is that this is only required on container rebuild, not restart, which can add a lot of start up delays. As well as apt prospectively updating components out of band.

The current solution with the `startup.sh` script also adds some delay to the container startup, but extensions can be updated without re-building or re-creating the container, and would require re-install of the dependencies, which is why we have it as a startup script.

**Comment by simonmcnair** [comment]
This is my latest since the image changes were implemented in https://github.com/AbdBarho/stable-diffusion-webui-docker/releases/tag/7.0.0:
```
pip install --upgrade pip

pip install diffusers discord_webhook trimesh rich vispy numexpr moviepy==1.0.2 networkx==2.5 imageio imageio_ffmpeg transforms3d==0.4.1  importlib-metadata importlib-metadata matplotlib numba==0.57.0 rembg protobuf==3.20.0 send2trash dynamicprompts
pip install -q --force-reinstall opencv-python-headless
```

**Comment by simonmcnair** [comment]
I can confirm some of depth map will not work without cuda-libraries-dev-11-8 installed.  that results in the EGL error
